### PR TITLE
Added Nette\Utils\Floats helper

### PR DESCRIPTION
**Pull request by mesour**
- new feature
- BC break? no
- doc PR: https://github.com/nette/docs/pull/848

---

Added support for comparing floats

The original class was created by @fabik.

I've added tests, and some fixes in the method `compare`.

### CODE REVIEW

File: src/Utils/Numbers.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+namespace Nette\Utils;
+
+use Nette;
+
+
+/**
+ * Floats tools library.
+ */
+class Numbers
```
Comment (JanTvrdik) [inline src/Utils/Numbers.php:18]: Should be called `Floats`.

File: src/Utils/Numbers.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+	public static function isZero(float $value): bool
+	{
+		return abs($value) < self::$epsilon;
+	}
+
+
+	public static function isInteger(float $value): bool
+	{
+		return abs(floor($value) - $value) < self::$epsilon;
```
Comment (JanTvrdik) [inline src/Utils/Numbers.php:34]: imho `round` must be used instead of `floor`

File: src/Utils/Numbers.php [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+/**
+ * Floats tools library.
+ */
+class Numbers
+{
+	use Nette\StaticClass;
+
+	/** @var float */
+	private static $epsilon = 1e-9;
```
Comment (JanTvrdik) [inline src/Utils/Numbers.php:23]: should be constant

File: src/Utils/Numbers.php [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+	public static function isInteger(float $value): bool
+	{
+		return abs(floor($value) - $value) < self::$epsilon;
+	}
+
+
+	public static function divide(float $dividend, float $divisor): float
+	{
+		if (self::isZero($divisor)) {
```
Comment (JanTvrdik) [inline src/Utils/Numbers.php:40]: Not sure how useful this is. Division by zero is well defined for floats. If you don't want to get `+Inf`, `-Inf` or `Nan` as a result, you would need to check more things.

File: src/Utils/Numbers.php [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+			}
+			throw new DivisionByZeroException('Trying to get a percentage of the total of 0.');
+		}
+		return ($part / $total) * 100;
+	}
+
+
+	public static function compare(float $a, float $b): int
+	{
+		if (abs($a - $b) < self::$epsilon) {
```
Comment (JanTvrdik) [inline src/Utils/Numbers.php:62]: It's not that simple https://github.com/nette/tester/pull/68

File: src/Utils/Numbers.php (hunk 2 above)
Comment (mesour) [reply src/Utils/Numbers.php]: agree

File: src/Utils/Numbers.php (hunk 1 above)
Comment (mesour) [reply src/Utils/Numbers.php]: agree

File: src/Utils/Numbers.php (hunk 4 above)
Comment (mesour) [reply src/Utils/Numbers.php]: I don't know how much, but I did not encounter this problem. And I think, here is better to check at least zero than nothing.

If you have an example that would return NAN or INF, feel free to send it and try to fix it in that method.

File: src/Utils/Numbers.php (hunk 5 above)
Comment (mesour) [reply src/Utils/Numbers.php]: Upgraded. Now it works too with INF and NAN.

File: src/Utils/Numbers.php (hunk 4 above)
Comment (JanTvrdik) [reply src/Utils/Numbers.php]: * `Nan / 3.0`
* `Inf` / 4.0

File: src/Utils/Floats.php [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+	private const EPSILON = 1e-10;
+
+
+	public static function isZero(float $value): bool
+	{
+		return abs($value) < self::EPSILON;
+	}
+
+
+	public static function isInteger(float $value): bool
```
Comment (janbarasek) [inline src/Utils/Floats.php:32]: Epsilon should be user-defined. :)

Comment (mesour) [reply src/Utils/Floats.php:32]: Now it is public static property :)

Comment (mesour) [reply src/Utils/Floats.php:32]: Changed back to constant. If it turns out why it should be user-defined, then we make an adjustment.

**Comment by dg** [comment]
Great, thanks

**Pull request merged**